### PR TITLE
parsers: Implement basic Cadence Insicive Enterprise Simulator parser

### PR DESCRIPTION
**Pull request by nekromant**
This simple patch implements basic parsing of Cadence HDL Simulation tools. multi-line warnings/error aren't supported and it relies on GNU/Make to print "entering directory" when constructing file paths. 

The relevant test is also enclosed. 

Signed-off-by: Andrew Andrianov <andrew@ncrmnt.org>

### CODE REVIEW

**Comment by uhafner** [comment]
Uups, seems that I missed that PR during my vacation. Thanks for the PR!

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    public CadenceIncisiveParser() {
+        super(Messages._Warnings_CadenceIncisive_ParserName(),
+                Messages._Warnings_CadenceIncisive_LinkName(),
+                Messages._Warnings_CadenceIncisive_TrendName(),
+                CADENCE_MESSAGE_PATTERN);
+    }
+
+    @Override
+    protected String getId() {
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Remove: getId makes only sense if you change the IDs.

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            /* column = matcher.group(14); */
+            message = matcher.group(15);
+        } else if (matcher.group(21) != null) {
+            tool = matcher.group(22);
+            type = matcher.group(23);
+            category = matcher.group(24);
+            message = matcher.group(25);            
+            fileName = "/NotFileRelated";
+        } else {
+            return FALSE_POSITIVE; /* Some shit happened ! */
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Comment: should never happen

File: src/test/java/hudson/plugins/warnings/parser/CadenceIncisiveParserTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Tests the class {@link CadenceIncisiveParserTest}.
+ *
+ * @author Andrew 'Necromant' Andrianov
+ */
+public class CadenceIncisiveParserTest extends ParserTester {
+    private static final String WARNING_TYPE = new CadenceIncisiveParser().getGroup();
+
+    /**
+     * Test of createWarning method, of class {@link GnuMakeGccParser}.
```
Comment (uhafner) [inline src/test/java/hudson/plugins/warnings/parser/CadenceIncisiveParserTest.java]: Typo.

File: src/test/resources/hudson/plugins/warnings/parser/CadenceIncisive.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+make[2]: Entering directory '/tmp/build-dir'
+ncelab: *W,CUSRCH: Resolved design unit 'shittyram' at 'u_shittyrams' to 'shittysoc.shittyram:v' through a global search of all libraries.
```
Comment (uhafner) [inline src/test/resources/hudson/plugins/warnings/parser/CadenceIncisive.txt]: It would be nice if your names would not use such a rude language.

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    private Warning handleDirectory(final Matcher matcher, int offset) {
+        directory = matcher.group(offset) + SLASH; //17
+        return FALSE_POSITIVE;
+    }
+
+    
+    @Override
+    protected Warning createWarning(final Matcher matcher) {
+
+        String tool, type, category, message, fileName;
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Single line per variable. Also move variables to the position where they are used first.

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    protected Warning createWarning(final Matcher matcher) {
+
+        String tool, type, category, message, fileName; 
+        int lineNumber = 0;
+        int column = 0;
+
+
+        List<String> arr = new ArrayList<String>();
+        int n = matcher.groupCount();
+        for (int i = 0; i<= n; i++)
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Braces!

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            arr.add(i,matcher.group(i));
+        
+        Priority priority = Priority.LOW;
+        
+        if (matcher.group(1) != null)
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Please reformat the next lines to the end using IntelliJ (or Eclipse) formatter.

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            tool = matcher.group(22);
+            type = matcher.group(23);
+            category = matcher.group(24);
+            message = matcher.group(25);            
+            fileName = "/NotFileRelated";
+        } else {
+            return FALSE_POSITIVE; /* Some shit happened ! */
+        }
+                 
+        if (category.equalsIgnoreCase("E")) {
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Is category always not null?

File: src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            priority = Priority.HIGH;
+            category = "Error (" + tool + "): " + category;
+        } else {
+            category = "Warning (" + tool + "): " + category;
+        }
+
+        if (fileName.startsWith(SLASH)) {
```
Comment (uhafner) [inline src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java]: Is filename always not null?

File: src/test/java/hudson/plugins/warnings/parser/CadenceIncisiveParserTest.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Collection;
+import java.util.Iterator;
+
+import org.junit.Test;
+
+import hudson.plugins.analysis.util.model.FileAnnotation;
+import hudson.plugins.analysis.util.model.Priority;
+
+/**
+ * Tests the class {@link CadenceIncisiveParserTest}.
```
Comment (uhafner) [inline src/test/java/hudson/plugins/warnings/parser/CadenceIncisiveParserTest.java]: CadenceIncisiveParser

**Comment by nekromant** [comment]
Thanks for the review, I hope I'll get to fix it the following week. Is it okay to just add the fixes in new commits or would you prefer changes melded into the current commits?

**Comment by uhafner** [comment]
Just add new commits.

**Comment by nekromant** [comment]
@uhafner Sorry for the delay, I've fixed everything you've noted.

**Comment by nekromant** [comment]
Rebased, so that it can be merged with no problem

**Comment by nekromant** [comment]
@uhafner Updated so that CI checks pass.

**Comment by uhafner** [comment]
While migrating the parsers to a new API I found some problems in your parser. Are you still available for some questions or fixes?

**Comment by nekromant** [comment]
Yes, sure.

**Comment by uhafner** [comment]
In [line 103 ](https://github.com/jenkinsci/warnings-plugin/blob/master/src/main/java/hudson/plugins/warnings/parser/CadenceIncisiveParser.java#L103) you are matching the category with a single character. Shouldn't this be a check for the type of the warning? (Otherwise the type will not be used at all.) There is no test for this part so I'm not sure which approach is correct... 

The new changed code is now here (new library): 
https://github.com/jenkinsci/analysis-model/commit/f6365e7237dc06dc83878b6165582fbfb525d704#diff-3e67025ce08b38593ffe3ef1fdaed513R101

Do you think that this is ok?

**Comment by nekromant** [comment]
Looks good to me.

As far as I remember, that's due to the fact that cadence has different warning (and error) printout styles for different tools from the suite. In some corner case "E" would end up in category variable, so I added that little hack. 

I can't tell anything more precise or test it, since I'll only be able to VPN to work and give the new version of the plugin a try on real-life logs only somewhere around the first days of 2018.
